txpool/txpool.go

```diff
@@ -192,576 +200,280 @@ func NewTxPool(
-		t.logger.Debug(
-			fmt.Sprintf(
-				"Rejecting tx [%s] from account heap due to low nonce",
-				tx.Hash.String()),
-		)
-
-		return ErrNonceTooLow
-	}
-
-	// check for slot overflow and handle accordingly
-	if err := t.processSlots(tx, isLocal); err != nil {
-		return err
-	}
 
-	t.logger.Debug("add txn", "ctx", origin, "hash", tx.Hash, "from", tx.From)
-
-	wrapper := t.lockAccountQueue(tx.From, true)
-	defer wrapper.unlock()
-
-	wrapper.accountQueue.Add(tx)
-
-	if !isLocal {
-		if pushErr := t.remoteTxns.Push(tx); pushErr != nil {
-			t.logger.Error(
-				fmt.Sprintf(
-					"Unable to push txn [%s] to the remote txns queue",
-					tx.Hash.String(),
-				),
-			)
-		}
-	}
-
-	// Skip check of GasPrice in the future transactions created by same address when TxPool receives transaction by Gossip or Reorg
-	if isLocal && !t.locals.containsAddr(tx.From) {
-		t.locals.addAddr(tx.From)
-	}
-
-	// Move promotable txs to the pending queue
-	for _, promoted := range wrapper.accountQueue.Promote() {
-		if pushErr := t.pendingQueue.Push(promoted); pushErr != nil {
-			t.logger.Error(fmt.Sprintf("Unable to promote transaction %s, %v", promoted.Hash.String(), pushErr))
-		} else {
-			t.metrics.PendingTxs.Add(1)
+		if err := p.topic.Publish(tx); err != nil {
+			p.logger.Error("failed to topic tx", "err", err)
 		}
 	}
 
-	// Add the transaction to the lookup map
-	t.addTxToLookup(tx)
-
 	return nil
 }
 
-// DecreaseAccountNonce resets the nonce attached to an account whenever a transaction produce an error which is not
-// recoverable, meaning the transaction will be discarded.
-//
-// Since any discarded transaction should not affect the world state, the nextNonce should be reset to the value
-// it was set to before the transaction appeared.
-func (t *TxPool) DecreaseAccountNonce(tx *types.Transaction) {
-	wrapper := t.lockAccountQueue(tx.From, true)
-	defer wrapper.unlock()
-
-	if wrapper.accountQueue.nextNonce > 0 {
-		wrapper.accountQueue.nextNonce--
+// Generates all the transactions (primaries)
+// ready for execution.
+func (p *TxPool) Prepare() {
+	// clear from previous round
+	if p.executables.length() != 0 {
+		p.executables.clear()
 	}
-}
 
-// GetTxs gets pending and queued transactions
-func (t *TxPool) GetTxs(inclQueued bool) (map[types.Address]map[uint64]*types.Transaction, map[types.Address]map[uint64]*types.Transaction) {
-	t.pendingQueue.lock.Lock()
-	pendingTxs := make(map[types.Address]map[uint64]*types.Transaction)
-	sortedPricedTxs := t.pendingQueue.index
-	for _, sortedPricedTx := range sortedPricedTxs {
-		if _, ok := pendingTxs[sortedPricedTx.from]; !ok {
-			pendingTxs[sortedPricedTx.from] = make(map[uint64]*types.Transaction)
-		}
-		pendingTxs[sortedPricedTx.from][sortedPricedTx.tx.Nonce] = sortedPricedTx.tx
-	}
-	t.pendingQueue.lock.Unlock()
-	if !inclQueued {
-		return pendingTxs, nil
-	}
+	// fetch primary from each account
+	primaries := p.accounts.getPrimaries()
```

Comment:
Perhaps nextInLine or head might be more descriptive than primary?

Reply:
Not sure what the ideal name for this method is, but `getHeads` seems even more vague.

Perhaps a detailed description on what `primaries` are ?